src/main/java/org/truffleruby/collections/WeakValueCache.java

```diff
@@ -56,9 +58,30 @@ public final Value get(Key key) {
         return reference.get();
     }
 
-    public final void put(Key key, Value value) {
+    /** Returns the value in the cache (existing or added) */
+    public Value addInCacheIfAbsent(Key key, Value newValue) {
         removeStaleEntries();
-        map.put(key, new KeyedReference<>(value, key, deadRefs));
+
+        final KeyedReference<Key, Value> newRef = new KeyedReference<>(newValue, key, deadRefs);
+
+        while (true) {
```

Comment:
What algorithm is this loop? Can you comment in the file about what the idea of this code is. Is it a loop to keep trying to modify the structure? Retrying if there is a conflict?

Reply:
It's a traditional CAS loop but on a Map instead of an atomic variable.
This operation is a `putIfAbsent()` but named differently as it returns the value in the map, no matter if it was already there or we had to insert it.
In addition to a `putIfAbsent()`, it is also checking if the entry in the map is stale, in which case we must replace it.
I'll add a comment about that.

Reply:
Comment added.